crates/biome_js_formatter/src/context.rs

```diff
@@ -44,12 +45,19 @@ pub struct JsFormatContext {
     cached_function_body: Option<(AnyJsFunctionBody, FormatElement)>,
 
     source_map: Option<TransformSourceMap>,
+
+    foreign_language_formatter: Rc<dyn JsForeignLanguageFormatter>,
```

Comment:
I'm just curious why Rc<dyn> is a better choice over other options?

Reply:
I didn't think about it deeply and just followed the `comments`. Any better suggestions?

Reply:
Yeah, I also noticed the `comments` and couldn't find an explanation.
Unfortunately, I don't have any suggestions about it.
There are several options:

- Have a `<T>` and have a constraint over `T` in the impls.
- `Box<dyn>`?

Reply:
Oh, sorry, you mean `Rc<dyn>`. I thought you were asking why use `Rc` instead of another smart pointer.

Initially, I wanted to use a closure, but I encountered a problem with deriving the `Clone` trait, so I chose to use Rc<dyn> instead.

Reply:
> There are several options:

> Have a <T> and have a constraint over T in the impls.
> Box<dyn>?

Thank you! I will think about it!

Reply:
Oh, I got it. Indeed, if we need the `Clone` trait, we can't use `Box` :(
It may be fine to use `Rc`.

Reply:
It seems `dyn` has [trade-offs](https://doc.rust-lang.org/nightly/std/keyword.dyn.html#trade-offs). still need to think about it, convert this PR to a draft temporarily

Reply:
> Have a T and have a constraint over T in the impls.

If we attach a generic type to `JsFormatContext`, we also need to attach a generic type to many other types associated with `JsFormatContext`, for example  [FormatJsSyntaxToken](https://github.com/biomejs/biome/blob/08f0c8f1078d31ec5764532dccfeec84146560f1/crates/biome_js_formatter/src/lib.rs#L428C17-L428C36). 

The [rust book](https://doc.rust-lang.org/book/ch17-02-trait-objects.html#trait-objects-perform-dynamic-dispatch) mentions that `dyn` has runtime cost but I have no idea how much the cost it is. Wonder if the cost is acceptable while keeping the code flexible

Reply:
hopefully you can fix it @ah-yu

Reply:
> The [rust book](https://doc.rust-lang.org/book/ch17-02-trait-objects.html#trait-objects-perform-dynamic-dispatch) mentions that dyn has runtime cost but I have no idea how much the cost it is. Wonder if the cost is acceptable while keeping the code flexible

It looks like the overhead will be totally acceptable in this case, so I’d just stick with the `Rc<dyn>`. After all, it’s a single pointer to the formatter that should be used, but it shouldn’t affect the performance of the formatter itself.

Generally in Rust, don’t worry too much about performance unless you’re making clones all over the place or something is becoming noticeably slow :)